include/aws/common/atomics_msvc.inc

```diff
@@ -0,0 +1,324 @@
+ * order_failure must be no stronger than order_success, and must not be release or acq_rel.
+ */
+AWS_STATIC_IMPL
+bool aws_atomic_compare_exchange_ptr(
+    volatile struct aws_atomic_var *var,
+    void **expected,
+    void *desired,
+    enum aws_memory_order order_success,
+    enum aws_memory_order order_failure) {
+    aws_atomic_priv_check_order(order_success);
+    aws_atomic_priv_check_order(order_failure);
+
+    void *oldval = InterlockedCompareExchangePointer(&var->ptrval, desired, *expected);
+    bool successful = oldval == *expected;
+    *expected = oldval;
+
+    return successful;
+}
+
+/**
+ * Atomically adds n to *var, and returns the previous value of *var.
+ */
+AWS_STATIC_IMPL
+aws_atomic_int_impl_t aws_atomic_fetch_add(
+    volatile struct aws_atomic_var *var,
+    aws_atomic_int_impl_t n,
+    enum aws_memory_order order) {
+    aws_atomic_priv_check_order(order);
+
+    return InterlockedExchangeAdd64(&var->intval, n);
+}
+
+/**
+ * Atomically subtracts n from *var, and returns the previous value of *var.
+ */
+AWS_STATIC_IMPL
+aws_atomic_int_impl_t aws_atomic_fetch_sub(
+    volatile struct aws_atomic_var *var,
+    aws_atomic_int_impl_t n,
+    enum aws_memory_order order) {
+    aws_atomic_priv_check_order(order);
+
+    return InterlockedExchangeAdd64(&var->intval, -n);
+}
+
+/**
+ * Atomically ORs n with *var, and returns the previous value of *var.
+ */
+AWS_STATIC_IMPL
+aws_atomic_int_impl_t aws_atomic_fetch_or(
+    volatile struct aws_atomic_var *var,
+    aws_atomic_int_impl_t n,
+    enum aws_memory_order order) {
+    aws_atomic_priv_check_order(order);
+
+    return InterlockedOr64(&var->intval, n);
+}
+
+/**
+ * Atomically ANDs n with *var, and returns the previous value of *var.
+ */
+AWS_STATIC_IMPL
+aws_atomic_int_impl_t aws_atomic_fetch_and(
+    volatile struct aws_atomic_var *var,
+    aws_atomic_int_impl_t n,
+    enum aws_memory_order order) {
+    aws_atomic_priv_check_order(order);
+
+    return InterlockedAnd64(&var->intval, n);
+}
+
+/**
+ * Atomically XORs n with *var, and returns the previous value of *var.
+ */
+AWS_STATIC_IMPL
+aws_atomic_int_impl_t aws_atomic_fetch_xor(
+    volatile struct aws_atomic_var *var,
+    aws_atomic_int_impl_t n,
+    enum aws_memory_order order) {
+    aws_atomic_priv_check_order(order);
+
+    return InterlockedXor64(&var->intval, n);
+}
+
+/**
+ * Provides the same reordering guarantees as an atomic operation with the specified memory order, without
+ * needing to actually perform an atomic operation.
+ */
+AWS_STATIC_IMPL
+void aws_atomic_thread_fence(enum aws_memory_order order) {
+    volatile int x = 0;
+    aws_atomic_priv_check_order(order);
+
+    switch (order) {
+        case aws_memory_order_seq_cst:
+            MemoryBarrier();
+            break;
+        case aws_memory_order_release:
+        case aws_memory_order_acquire:
+        case aws_memory_order_acq_rel:
+            x = x;
```

Comment:
Weirdly, MSVC's thread_fence only uses MemoryBarrier() for ARM, and for other architectures uses an atomic exchange on a dummy static.

```
inline void _Atomic_thread_fence(memory_order _Order)
	{	/* force memory visibility and inhibit compiler reordering */
 #if defined(_M_ARM) || defined(_M_ARM64)
	if (_Order != memory_order_relaxed)
		{
		_Memory_barrier();
		}

 #else
	_Compiler_barrier();
	if (_Order == memory_order_seq_cst)
		{	/* force visibility */
		static _Uint4_t _Guard;
		_Atomic_exchange_4(&_Guard, 0, memory_order_seq_cst);
		_Compiler_barrier();
		}
 #endif
	}
```

just pointing out the differences